Project: Implement Benchmark Test Cases

ericpolo opened the pull request.

ericpolo committed:
Project: Implement Benchmark Test Case

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+
+int FindTableIndex(string tableName)
+{
+    for (int i = 0; i < NUM_TABLES; i++) {
+        if (tableNames[i] == tableName) {
+            return i;
+        }
+    }
+
+    KU_ASSERT(0);
+    return 0;
```
Optional: Can be replaced with the following:
```
string *end = tableNames + sizeOf(tableNames)/sizeOf(tableNames[0]);
int* it = std::find(tableNames, end, tableName);
KU_ASSERT(it != end);
return *it;
```

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+        "abcdefghijklmnopqrstuvwxyz";
+    std::string tmp_s;
+    tmp_s.reserve(len);
+
```
Consider specifying a seed for the random number generator, unless we want the same random numbers to be generated (hence same random strings, tests etc.) in every run of the program:
```
// Additional imports that may be required for the seed:
#include <cstdlib> // for rand() and srand()
#include <ctime>   // for time()

// Seed the random number generator.
srand(static_cast<unsigned int>(time(0)));
```

Additionally, is thread safety a requirement? If yes, consider using `<random>`.

ericpolo replied:
Fixed in the newest comment. we only have one thread for the test, so thread safety is unnecessary.

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+#include <iostream>
+#include <string>
+#include "build/debug/src/kuzu.hpp"
```
Are we planning to always run our tests on debug mode? Perhaps we should use conditional imports here based on a compile flag argument, something as follows:
```
#ifdef DEBUG
#include "build/debug/src/kuzu.hpp"
#else
#include "build/release/src/kuzu.hpp
#endif"
```
To use the debug flag, the compilation would require the flag, something of the following form:
`g++ -DDEBUG benchmark.cpp`

ericpolo replied:
Fixed in the newest commit. Also updated the README file

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+
+string RandomSelectTable(string avoidTable)
+{
+    string tableName = avoidTable;
+    while (tableName == avoidTable) {
+        int tableIndex = rand() % NUM_TABLES;
+        tableName = tableNames[tableIndex];
+    }
```
Optional - do-while loop is better suited for this.

ericpolo replied:
Fixed

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+
+void CreateTable(const unique_ptr<Connection> &connection, string tableName)
+{
+    for (auto i = 0; i < NUM_TABLES; i++) {
+        auto name = tableNames[i];
+        if (tableName == name) {
+            connection->query(tableCreateQuery[i]);
+            connection->query(tableCopyQuery[i]);
+        }
+    }
```
Minor - it seems we can directly use the `findTableIndex` function here:
```
int idx = findTableIndex(tableName);
connection->query(tableCreateQuery[idx]);
connection->query(tableCopyQuery[idx]);
```

ericpolo replied:
Fixed in the newest commit.

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+
+char* GetCmdOption(char ** begin, char ** end, const std::string & option)
+{
+    char ** itr = std::find(begin, end, option);
```
Minor - `std::` is probably redundant since `using namespace std;` is already specified.

ericpolo replied:
Fixed

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+    return nullptr;
+}
+
+void UpdateTableCopyQuery(string csvFileDir)
```
Recommend passing strings as const references wherever possible, to avoid overhead of copying string values:
```
void updateTableCopyQuery(const string& csvFileDir);
```
Applicable throughout the file.

ericpolo replied:
Fixed in the newest commit

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+}
+
+/* Note that colType is only valid when alterType is ADD_COLUMN. Otherwise, it should be set as INVALID_COLUMN_TYPE */
+void AlterTable(const unique_ptr<Connection> &connection, string tableName, AlterType alterType, string colName, AlterAddColumnType colType)
```
Going by the comment, consider specifying a default value for colType (provided the C++ version supports it), so that no value needs to be specified at the caller side if the alter type is not ADD_COLUMN:

```
void alterTable(const unique_ptr<Connection> &connection, string tableName, AlterType alterType, string colName, AlterAddColumnType colType = INVALID_COLUMN_TYPE) {
// existing code
}
```
Now the function may be called as follows:
```
// DROP COLUMN example
 AlterTable(connection, tableName, DROP_COLUMN, dropColName);
// ADD_COLUMN_EXAMPLE
 AlterTable(connection, tableName, ADD_COLUMN, addColName, INT64);
```

ericpolo replied:
Fixed in the newest commit

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+
+int main(int argc, char* argv[])
+{
+    if (argc == 1) {
```
Since we are expecting at least 3 arguments, the condition may be modified accordingly.

MayankJasoria replied:
4 now with the updated code

ericpolo replied:
Fixed. We actually need 7 since argc will also count "-N/-B/-D"

MayankJasoria commented on benchmark848/benchmark.cpp:
```diff
@@ -0,0 +1,348 @@
+                    break;
+                }
+                case ALTER_TABLE: {
+                    /* ERICTODO: Calculate average checkpoint time here */
```
In case this helps:

The checkpoint function could be modified to return the running time, which we can then use to compute average times. Consider the following:
```
#include <chrono>

std::chrono::nanoseconds checkpoint(const unique_ptr<Connection> &connection) {
    // Get the starting timepoint
    auto start = std::chrono::high_resolution_clock::now();

    // run the checkpoint query
    connection->query("CHECKPOINT;");

    // Get the ending timepoint
    auto end = std::chrono::high_resolution_clock::now();

    // Calculate the duration in nanoseconds
    auto duration_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start);

   return duration_ns;
}
```
The duration_ns may be printed or arithmetic operations may be performed, using which we should be able to estimate average runtime.

ericpolo committed:
Project: Improve benchmark test cases
  1. Add stat collection to benchmark
  2. Add macro to control FCM availability
  3. Solve the bug that iterate forever in getFreeChunk when failed to
     find a chunk at the current FreeChunkLevel
  4. Update FreeChunkLevel to better suit the need of test cases

ericpolo committed:
Project: Fix Mayank's comments
